### PR TITLE
Add EVAL_RO

### PR DESCRIPTION
**Pull request by xiaobiaozhao**
This is design https://github.com/apache/incubator-kvrocks/discussions/720

useage
`127.0.0.1:6666> EVAL_RO "redis.call('SET', 'a', 'bb') return {ARGV[1], redis.read_only}" 0 hello
(error) ERR running script (call to f_963e3d4d3ef14756826239b888ba1c2257052a58): @user_script:1: Write commands are not allowed from read-only scripts`

### CODE REVIEW

File: tests/tcl/tests/unit/scripting.tcl [hunk 1]
```diff
@@ -387,6 +387,17 @@ start_server {tags {"scripting"}} {
...
+    test {EVAL_RO - Successful case} {
+        r set foo bar
+        assert_equal bar [r eval_ro {return redis.call('get', KEYS[1]);} 1 foo]
+    }
+
+    test {EVAL_RO - Cannot run write commands} {
+        r set foo bar
+        catch {r eval_ro {redis.call('del', KEYS[1]);} 1 foo} e
+        set e
+    } {ERR Write commands are not allowed from read-only scripts*}
```
Comment (PragmaTwice) [inline tests/tcl/tests/unit/scripting.tcl]: Consider to add test for `evalsha_ro`?

File: src/redis_cmd.cc [hunk 2]
```diff
@@ -4742,6 +4742,32 @@ class CommandEvalSHA : public Commander {
   }
 };
 
+class CommandEvalRO : public Commander {
+ public:
+  Status Parse(const std::vector<std::string> &args) override {
+    return Status::OK();
+  }
+
```
Comment (git-hulk) [inline src/redis_cmd.cc]:
```suggestion
```
Can remove this part if it does nothing

File: src/scripting.cc [hunk 3]
```diff
@@ -898,7 +921,9 @@ Status createFunction(Server *srv, const std::string &body, std::string *sha) {
   funcdef += body;
   funcdef += "\nend";
 
-  lua_State *lua = srv->Lua();
+  if (!lua) {
```
Comment (git-hulk) [inline src/scripting.cc]: Can we choose the Lua state in caller? the caller should determine where to create the function.

File: src/scripting.cc [hunk 4]
```diff
@@ -233,11 +240,18 @@ namespace Lua {
...
     char funcname[43];
     Server *srv = conn->GetServer();
-    lua_State *lua = srv->Lua();
+    lua_State *lua = NULL;
+    if (read_only) {
+      // use worker's lua VM
+      lua = conn->Owner()->Lua();
+    } else {
+      lua = srv->Lua();
+    }
```
Comment (git-hulk) [inline src/scripting.cc]:
```suggestion
    lua_State *lua = srv->Lua();
    if (read_only) {
      // Use the worker's private Lua VM when entering the read-only mode
      lua = conn->Owner()->Lua();
    }
```

**Comment by git-hulk** [comment]
Another issue is phantom read in read-only mode but it can't be solved it in this PR: 

For example:

```
Time      Worker A       Worker B    

 T1         Get K0  

 T2                       Set K0 V0'

 T3         Get K0
```

File: src/redis_cmd.cc (hunk 2 above)
Comment (datavisorxiaobiaozhao) [reply src/redis_cmd.cc]: done

File: tests/tcl/tests/unit/scripting.tcl (hunk 1 above)
Comment (xiaobiaozhao) [reply tests/tcl/tests/unit/scripting.tcl]: added

File: src/scripting.cc (hunk 4 above)
Comment (xiaobiaozhao) [reply src/scripting.cc]: done

File: src/scripting.cc (hunk 3 above)
Comment (xiaobiaozhao) [reply src/scripting.cc]: done

**Comment by tisonkun** [comment]
@xiaobiaozhao would you like to rebase this patch onto latest unstable? Then I can find some time to review it :)

**Comment by git-hulk** [comment]
@xiaobiaozhao Can help to resolve conflicts and take a look at why the CI NOT works.

**Comment by xiaobiaozhao** [comment]
> 

done

**Comment by xiaobiaozhao** [comment]
> @xiaobiaozhao Can help to resolve conflicts and take a look at why the CI NOT works.
done

**Comment by git-hulk** [comment]
Thanks, will take a look soon.

**Review by git-hulk — DISMISSED** [review]

**Review by tisonkun — DISMISSED** [review]

**Comment by git-hulk** [comment]
@ShooterIT @PragmaTwice @caipengbo This PR will be merged in tomorrow if we have no more comments.

**Comment by PragmaTwice** [comment]
> Another issue is phantom read in read-only mode but it can't be solved it in this PR:
> 
> For example:
> 
> ```
> Time      Worker A       Worker B    
> 
>  T1         Get K0  
> 
>  T2                       Set K0 V0'
> 
>  T3         Get K0
> ```

Has this issue been resolved now?

**Comment by tisonkun** [comment]
@git-hulk this patch has a code conflict with #900. But I can simply update after this patch merged :)

**Comment by tisonkun** [comment]
@PragmaTwice this manner seems by design for me O_O

Generally, Kvrocks doesn't employ a DBMS ACID model so "phantom read" is not well defined in this scope.

**Comment by git-hulk** [comment]
> > Another issue is phantom read in read-only mode but it can't be solved it in this PR:
> > For example:
> > ```
> > Time      Worker A       Worker B    
> > 
> >  T1         Get K0  
> > 
> >  T2                       Set K0 V0'
> > 
> >  T3         Get K0
> > ```
> 
> Has this issue been resolved now?

This issue is NOT introduced by this PR and can't be resolved in short time. We need to support  the version read mechanism in all readonly commands.

**Review by PragmaTwice — DISMISSED** [review]

**Review by caipengbo — DISMISSED** [review]

**Comment by git-hulk** [comment]
> @git-hulk this patch has a code conflict with #900. But I can simply update after this patch merged :)

Okay, thanks. I think we can remove the command count test case since it's meaningless.

**Comment by PragmaTwice** [comment]
> @PragmaTwice this manner seems by design for me O_O
> 
> Generally, Kvrocks doesn't employ a DBMS ACID model so "phantom read" is not well defined in this scope.

Got it. maybe I need to read the transaction chapter of DDIA again : (

**Comment by git-hulk** [comment]
> > @PragmaTwice this manner seems by design for me O_O
> > Generally, Kvrocks doesn't employ a DBMS ACID model so "phantom read" is not well defined in this scope.
> 
> Got it. maybe I need to read the transaction chapter of DDIA again : (

My bad. 😆 We have no ACID, so just behaves like phantom read.

**Review by git-hulk — APPROVED** [review]

**Comment by tisonkun** [comment]
@git-hulk @PragmaTwice the unstable go cases can be resolved by https://github.com/apache/incubator-kvrocks/pull/899. Please help with reviewing.

File: tests/gocase/unit/command/command_test.go [hunk 5]
```diff
@@ -35,13 +35,6 @@ func TestCommand(t *testing.T) {
 	rdb := srv.NewClient()
 	defer func() { require.NoError(t, rdb.Close()) }()
 
-	t.Run("Kvrocks supports 183 commands currently", func(t *testing.T) {
-		r := rdb.Do(ctx, "COMMAND", "COUNT")
-		v, err := r.Int()
-		require.NoError(t, err)
-		require.Equal(t, 183, v)
-	})
-
```
Comment (tisonkun) [inline tests/gocase/unit/command/command_test.go:44]: @git-hulk do we remove this test intentionally?

Comment (git-hulk) [reply tests/gocase/unit/command/command_test.go:44]: Yes, as said in comment: https://github.com/apache/incubator-kvrocks/pull/782#issuecomment-1252227937, this test case is not much help for us and it causes conflicts frequently, so I think we should remove.

Comment (tisonkun) [reply tests/gocase/unit/command/command_test.go:44]: Get it. Although, the code conflict doesn't relate to this test but (1) I'm porting scripting tests (2) this PR added new scripting tests XD.

**Pull request merged**